docs/protocol/off-chain/message-format.mdx

```diff
@@ -0,0 +1,433 @@
+    1699123457000
+  ],
+  "sig": [
+    "0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890ab"
+  ]
+}
+```
+
+---
+
+## Error Response
+
+When an error occurs, <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> sends an error message following the JSON-RPC 2.0 error format.
+
+### Structure
+
+```json
+{
+  "id": requestId,
+  "error": {
+    "code": errorCode,
+    "message": "Error description"
+  }
+}
+```
+
+### Error Codes
+
+Nitro RPC uses standard JSON-RPC error codes plus application-specific codes.
+
+#### Standard JSON-RPC Errors
+
+| Code | Name | Description |
+|------|------|-------------|
+| **-32700** | Parse error | Invalid JSON received by server |
+| **-32600** | Invalid request | Request structure is invalid |
+| **-32601** | Method not found | Requested method does not exist |
+| **-32602** | Invalid params | Method parameters are invalid |
+| **-32603** | Internal error | Internal server error |
+
+#### Application-Specific Errors (1000+)
+
+| Code | Name | Description |
+|------|------|-------------|
+| **1001** | Authentication required | Operation requires authenticated session |
+| **1002** | Insufficient balance | User doesn't have enough available funds |
+| **1003** | Channel not found | Specified channel_id does not exist |
+| **1004** | App session not found | Specified app_session_id does not exist |
+| **1005** | Quorum not met | Signatures don't meet quorum requirement |
+| **1006** | Version conflict | <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data">State</Tooltip> version doesn't match expected |
+| **1007** | Invalid allocation | Allocation sum doesn't match rules |
+| **1008** | Session key allowance exceeded | Operation would exceed spending cap |
+| **1009** | Unsupported chain | Specified chain_id not supported |
+| **1010** | Unsupported token | Specified token not supported |
+
+:::info Error Handling Best Practice
+Clients should implement proper error handling for both standard JSON-RPC errors (protocol-level) and application-specific errors (business logic-level). Always check the error code to determine appropriate recovery action.
+:::
+
+### Error Examples
+
+**Authentication Required**:
+```json
+{
+  "id": 5,
+  "error": {
+    "code": 1001,
+    "message": "Authentication required: session not established"
+  }
+}
+```
+
+**Insufficient Balance**:
+```json
+{
+  "id": 12,
+  "error": {
+    "code": 1002,
+    "message": "Insufficient balance: required 100 USDC, available 75 USDC"
+  }
+}
+```
+
+**Method Not Found**:
+```json
+{
+  "id": 8,
+  "error": {
+    "code": -32601,
+    "message": "Method not found: 'invalid_method'"
+  }
+}
+```
+
+---
+
+## State Hash Computation
```

Comment:
There is no such concept yet as off-chain state hash, instead, each RPC message (request or response) has a payload hash, payload is everything inside `req` or `res` field, we use _keccak256_ to hash a JSON-encoded payload (it should containing spaces only inside string values or after `,` and `:` signs in order to have a correct hash in result, - we may reconsider encoding logic in the next version of the protocol)

Reply:
I see that you mention a correct actual encoding logic in **Signature Verification Process** and **JSON Encoding Consistency** sections